In-plugin summary emails

truongwp opened the pull request:
**The Overview:** https://app.clickup.com/9006050445/v/dc/8ccuv4d-11611

The design for monthly and yearly email and the new settings: https://www.figma.com/file/jI9WfZElTkZIz8w9gAlVTu/Monthly%2C-Yearly-Review-Email?node-id=90%3A507&t=2dz6EarZGT3SB05Q-1 

**The technical pitch:** https://app.clickup.com/9006050445/v/dc/8ccuv4d-11571/8ccuv4d-5351

**Development tasks list:**
- [x] Create new settings
- [x] Add cron event
- [x] Add code to check when should send emails
- [x] Add base code for sending emails
- [x] Collect monthly data
- [x] Build monthly email content
- [x] Build yearly email content
- [x] Build license expired email content
- [ ] Unit tests

**Test cases:**
- [ ] There will be a new checkbox in the Misc section of Global settings, it should be enabled by default, and the Recipients setting displays. We will support `[admin_email]` in the Recipients setting.
- [ ] Make sure that no monthly or yearly email is sent in the first 15 days after updating.
- [ ] After 15 days, if it's less than 45 days to the renewal date, or the oldest form was created at least 1 year ago, a yearly email will be sent. Otherwise, a monthly email will be sent.
- [ ] Make sure that the time between 2 monthly or 1 monthly and 1 yearly email is at least 30 days.
- [ ] If the license is expired, a license expired email should be sent. And it's sent just once.
- [ ] Make sure emails are sent to the correct recipients (set in the new settings).
- [ ] To send email immediately, run the following code:

```
add_action( 'init', function() {
    if ( method_exists( 'FrmEmailSummaryHelper', 'send_monthly') ) {
        FrmEmailSummaryHelper::send_monthly(); // send monthly email for the last 30 days.
        FrmEmailSummaryHelper::send_yearly(); // send yearly email for the last 1 year.
    }
} );
```

Sent date tracking is in the option `frm_summary_emails_options`.

Use `add_filter( 'frm_html_summary_emails', '__return_false' );` to test the plain text emails.

**Note:** License expired email is moved to Formidable Pro. To test it, get the code in https://github.com/Strategy11/formidable-pro/pull/4594 then run `FrmProEmailLicenseExpiredController::send()`.

**QA site:**
I set up the QA site and added 3 snippets to send the summary emails. They are [here](https://qa.formidableforms.com/truong/wp-admin/admin.php?page=snippets), just run one of them once to test. Before running, make sure that you added your email address in `Formidable > Global settings > Miscellaneous > Recipients`.

Currently, I can't make the out-of-date plugins section included in the email. Look like it's related to multisite. I'm trying to make it work.

@stephywells You should review the content of those emails. Some URLs I'm not sure they are correct.
@tuguirazvan Please take a look at those emails and give me the design feedback.

truongwp commented:
@garretlaxton I think it will be hard for you to test all of this. I hope you can test the new settings to see if it works, and use my custom code to send the emails then test if their content is correct, the layout is good in different email clients, and they are sent to the right email address.

@Crabcyborg I requested your review. Please request any other people if needed. 

@stephywells Please check my test cases to see if I'm missing any cases.

For the design review, I'll push this to the QA site and set up some dummy data, so Razvan can check easily.

stephywells commented:
This is so great @truongwp ! This makes it SO much easier for everyone to know what to do, and cut down on questions. Thank you so much for the great information and details! 

Do we need to add anything to the test cases about checking that the stats included in the emails are correct? Are the payment details already in the email to work with Stripe Lite? I haven't tested it quite yet.

Crabcyborg commented on classes/helpers/FrmSummaryEmailsHelper.php:
```diff
@@ -0,0 +1,433 @@
+			$wpdb->prepare(
+				"SELECT fr.id AS form_id, fr.name AS form_name, COUNT(*) as items_count
+						FROM {$wpdb->prefix}frm_items AS it INNER JOIN {$wpdb->prefix}frm_forms AS fr ON it.form_id = fr.id
+						WHERE it.created_at >= %s AND it.created_at <= %s AND it.is_draft = 0
```
`it.created_at >= %s AND it.created_at <= %s` may be a bit more readable as `it.created_at BETWEEN %s AND %s` instead? I like `BETWEEN` as it's meant for this type of comparison on both sides.

Crabcyborg commented on classes/models/FrmSummaryEmail.php:
```diff
@@ -0,0 +1,78 @@
+		$subject    = $this->get_subject();
+		$headers    = $this->get_headers();
+
+		return wp_mail( $receptions, $subject, $content, $headers );
```
Normally in our emails we check if `wp_mail` failed and try to use `mail` as a fallback.

Is there any chance we can use some of that logic in `FrmEmail` so it's more consistent here?

truongwp replied:
I think we can move some code to a helper. Many methods in the `FrmEmail` are private so we can't inherit them.

truongwp replied:
I added the fallback. Do you think we need to reuse some logic of `FrmEmail`?

Crabcyborg commented on css/formidableforms1.css:
```diff
@@ -0,0 +1,2 @@
+/* WARNING: Any changes made to this file will be lost when your Formidable settings are updated */
```
It looks like this file was committed by mistake.

truongwp commented:
@stephywells I'll add some details for the stats data in the email. For the payment, I didn't add it. I just added some hooks to add any data to the Stats section.

@Crabcyborg Is Stripe in Lite stable enough for us to use it? Which branch is it on?

Crabcyborg commented:
@truongwp Stripe Lite is complete. It should be stable.

It's the stripe_lite branch.

garretlaxton commented:
Thanks Truong. 

`FrmSummaryEmailsHelper::send_license();` doesn't seem to work.

Also, is this link suppose to direct to somewhere? It currently just takes you to the homepage.
![image](https://github.com/Strategy11/formidable-forms/assets/16290508/7f0ca12f-3e1d-4326-ad4a-355eb913f619)

truongwp commented:
@stephywells Should we split or merge payments and subscriptions data?

stephywells commented:
@truongwp i think we can merge it all for now, and include a total of all revenue for the period.

truongwp commented:
@garretlaxton Sorry, I didn't update the correct method name in the code. Please test the `send_license()` again.

That URL should point to the home page.

garretlaxton commented:
The expired license renew now button in the email isn't clickable.
![image](https://github.com/Strategy11/formidable-forms/assets/16290508/dbc035b6-0298-4f59-9c33-c4d64f4376fd)

I also see the following warnings in my logs.
`PHP Warning:  Undefined property: stdClass::$response in /wp-content/plugins/formidable/classes/helpers/FrmSummaryEmailsHelper.php on line 433`

`PHP Warning:  foreach() argument must be of type array|object, null given in /wp-content/plugins/formidable/classes/helpers/FrmSummaryEmailsHelper.php on line 433`

truongwp commented:
@garretlaxton I fixed the PHP warning.

Which email client are you using? Can you forward the email with the unclickable Renew button to `truong@strategy11.com`?

garretlaxton commented:
I'm using gmail. I just forwarded the email to you.

truongwp commented:
@garretlaxton I fixed the problem with the renew link and added the payment data in the monthly and yearly emails. Please check them.

garretlaxton reviewed (approved):
Nice, this looks great for me now!
![image](https://github.com/Strategy11/formidable-forms/assets/16290508/fcc5b27b-a50e-44d5-9da0-2043952f41a4)
![image](https://github.com/Strategy11/formidable-forms/assets/16290508/bc7f9531-9e96-4866-964b-0c3cc887c14d)
![image](https://github.com/Strategy11/formidable-forms/assets/16290508/ee2ef2a2-1924-4ca9-8af5-8e622920040b)

Crabcyborg commented on classes/views/frm-settings/misc.php:
```diff
@@ -10,6 +10,25 @@
+			data-toggleclass="frm_summary_emails_recipients_wrapper"
+			<?php checked( $frm_settings->summary_emails, 1 ); ?>
+		/>
+		<?php esc_html_e( 'Allow Formidable Forms to send monthly and annual summary emails with your activity.', 'formidable' ); ?>
```
This setting name should probably only show "Formidable Forms" when the white labeling is off as well.

Crabcyborg commented on classes/models/FrmYearlyEmail.php:
```diff
@@ -0,0 +1,31 @@
+	}
+
+	protected function get_subject() {
+		return __( 'Your year with Formidable Forms', 'formidable' );
```
This line as well should probably conditionally include "Formdiable Forms".

truongwp replied:
@stephywells When using another white label, will we remove `Formidable Forms` or use that white label?

stephywells replied:
How about 'How your forms performed this year'. We could change this instead of doing something different for white label.

truongwp commented on classes/views/summary-emails/base-plain.php:
```diff
@@ -0,0 +1,21 @@
+?>
+%%INNER_CONTENT%%
+
+Keep conquering new heights with FormidableForms. Your progress fuels our passion!
```
There is a signature and the company detail under this line. Should we also remove them when `Formidable` isn't used in the white label setting?

Crabcyborg replied:
Before we start updating everything I'd like to know what @stephywells thinks. But I think we would leave out the signature and company detail when whitelabeling is on, yes.

Also, I'm noticing that it's FormidableForms here with no space. It should be two words.

stephywells replied:
That sounds like a good plan to leave out the signature if it's white labeled. Thanks for thinking about this Mike!

truongwp commented:
Thanks @Crabcyborg and @stephywells, I updated the code to handle the white label. Please take a look at the email content, in both HTML and plain text cases. I added a snippet on the QA site to disable HTML email.

Crabcyborg commented:
@truongwp I think this update looks great.

I think we should try to get some testing from the team on this.

@stephywells What do you think about scheduling a team QA session for the summary emails?

stephywells commented:
Awesome! That sounds great. This can be the QA for the meeting scheduled next week (or it might be in two weeks). Is that too far away? Since this isn't a huge feature, I think we'd be okay if we don't do the whole team. I would like to look at it before release either way. I'm not sure if I'll be able to join the next QA meeting or not.

stephywells commented on classes/models/FrmStatsEmail.php:
```diff
@@ -0,0 +1,168 @@
+	die( 'You are not allowed to call this page directly.' );
+}
+
+abstract class FrmStatsEmail extends FrmSummaryEmail {
```
This isn't a big deal, but it would be nice if the files all started the same so they'd be grouped in the list. If it would be easy, can we change these new classes to all start with "FrmEmail"? Like "FrmEmailStats", "FrmEmailSummary"...

stephywells replied:
Thank you! Looks like there are still a few of these like FrmSummaryEmailsHelper and FrmSummaryEmailsController

truongwp replied:
Ok. I renamed them.

stephywells replied:
Thank you!!

stephywells commented on classes/models/FrmLicenseExpiredEmail.php:
```diff
@@ -0,0 +1,34 @@
+	die( 'You are not allowed to call this page directly.' );
+}
+
+class FrmLicenseExpiredEmail extends FrmSummaryEmail {
```
Is it possible to move this email to pro?

truongwp replied:
Do you want to move all code related to license expired email or just this class? It's possible.

stephywells replied:
Yeah, I think that would be nice since it's not really related to lite.

truongwp replied:
I moved it to Pro. Please check https://github.com/Strategy11/formidable-pro/pull/4594

stephywells commented on classes/helpers/FrmSummaryEmailsHelper.php:
```diff
@@ -0,0 +1,474 @@
+	/**
+	 * Number of days to send the next monthly email.
+	 */
+	const MONTHLY_PERIOD = 3; // TODO: Change this to 30.
```
When are we planning on switching these numbers?

truongwp replied:
I updated this for easier to test. So I will restore it after we complete the testing.

stephywells replied:
Thanks! Just want to make sure we don't lose it.

stephywells reviewed (changes requested):
These emails and the code is looking really great! Sorry about the ridiculous delay in reviewing this!

Looks like we missed the "offer help" part in the designs. Let's add a one-liner like this:
https://www.figma.com/file/jI9WfZElTkZIz8w9gAlVTu/Monthly%2C-Yearly-Review-Email?node-id=1%3A607&mode=dev

stephywells commented on classes/helpers/FrmSummaryEmailsHelper.php:
```diff
@@ -0,0 +1,474 @@
+	 * @param int    $limit     Limit the result. Default is 10.
+	 * @return array            Contains `form_id`, `form_name`, and `items_count`.
+	 */
+	public static function get_top_forms( $from_date, $to_date, $limit = 10 ) {
```
I think this might be better to limit to 5 instead of 10.

truongwp commented on classes/helpers/FrmSummaryEmailsHelper.php:
```diff
@@ -0,0 +1,474 @@
+		}
+
+		if ( $diff > 0 ) {
+			$image = 'arrow-up.png';
```
What is your purpose? Do you want inline SVG or `.svg` file? Because the support for those type of SVGs is different.

stephywells replied:
I was thinking inline svgs since this is the only place we are using them.

truongwp replied:
It looks like there isn't much support for inline SVG https://www.caniemail.com/features/html-svg/
I think we shouldn't do that.

stephywells replied:
Okay, sounds good. In that case, let's update the arrow images. They feel fuzzy to me.

truongwp replied:
I replaced arrow icons with HTML special chars. It works great.

truongwp commented:
> These emails and the code is looking really great! Sorry about the ridiculous delay in reviewing this!
> 
> Looks like we missed the "offer help" part in the designs. Let's add a one-liner like this: https://www.figma.com/file/jI9WfZElTkZIz8w9gAlVTu/Monthly%2C-Yearly-Review-Email?node-id=1%3A607&mode=dev

Thanks for reviewing this. For the Offer help part, what is the link for the Contact support? Is it https://formidableforms.com/knowledgebase/support-2/?

truongwp commented on classes/views/summary-emails/license-expired-plain.php:
```diff
@@ -0,0 +1,25 @@
+
+echo "\r\n";
+
+esc_html_e( 'Your license is expired. You can go to the link below to renew.', 'formidable' );
```
If using striping HTML, some sections will be hard to manipulate, like Statistics or Top forms. I will create some methods for those strings to use in both types of email.

stephywells replied:
I was specifically looking at the renewal email. It looks like it would be helpful to keep the plain text version for the others.

truongwp replied:
Yes. I think we should keep the plain text versions. The texts in there are a bit different from the HTML versions.

stephywells replied:
Can we remove the plain version for the expired email?

truongwp replied:
Yes, we can. Can you give me the final content of that email?

stephywells replied:
Sure! Here it is:

> I hope you have loved the past year of using Formidable. It looks like your subscription has expired.
> Renewing your license grants you access to our legendary support services, form templates, Pro features, and updates for another year.
> [Renew Now] (Button) https://formidableforms.com/account/downloads/
> If you don't need Formidable Pro right now, that's ok. You can resubscribe in the future when needed.

truongwp replied:
Thank you. I updated the content and removed the plain text file for license email.

truongwp commented on classes/controllers/FrmHooksController.php:
```diff
@@ -107,6 +107,12 @@ public static function load_hooks() {
 		add_filter( 'frm_fields_in_form_builder', 'FrmFormsController::update_form_builder_fields', 10, 2 );
 
+		// Cronjob.
+		add_action( 'init', 'FrmCronController::schedule_events' );
```
The cron job could be missed if no request to the `wp-admin`.

stephywells replied:
Moving it would possibly delay the first schedule only, right? Once the schedule is set up, it shouldn't be a problem. Am I reading this right?

truongwp replied:
I tested this. `admin_init` works. The event still occurs when I load the frontend. Thank you.

truongwp replied:
But updating this will affect `formidable_send_usage`. Is it ok?

stephywells replied:
Looks like that one is scheduled with this line:
add_action( 'admin_init', 'FrmUsageController::schedule_send' );

But yes, that's fine to switch to admin_init

truongwp commented on classes/views/summary-emails/base.php:
```diff
@@ -0,0 +1,72 @@
+	<style type="text/css">
+		.frm_inbox_dismiss { display: none; }
+
+		.frm-button-primary {
```
This is for the CTA buttons. I updated the code to add inline CSS. But I think we should keep this as a fallback in case those types of buttons aren't processed.

stephywells replied:
Good plan. In that case, can we update it and add !important so it matches the white buttons?

truongwp replied:
Yes, we can. However, it may not work for all email clients.

stephywells replied:
Right. But if we keep it, it should match the styling we want.

stephywells replied:
Oh! I just realized why we aren't on the same page with this. I see you were shooting for a blue button, but I had the same white button in mind. I think since we don't have an image, the blue button feels a bit out of place.

Would you mind switching out the blue button styling on the inbox message to match the other white buttons in the email?

truongwp commented on classes/helpers/FrmSummaryEmailsHelper.php:
```diff
@@ -0,0 +1,474 @@
+	 * @param string $url The URL.
+	 * @return string
+	 */
+	public static function add_url_data( $url ) {
```
I updated the code to override it. Can you list the `utm_content` for each link?

stephywells replied:
Actually... Can we use the admin_upgrade_link() function? It looks like this will be creating duplication.

We can use values like 'unsubscribe_url', 'site_url', 'plugins_url'... These are the names that are being used for them in the code.

truongwp replied:
I think we can't use `admin_upgrade_link()` without modifying it. It prepends `https://formidableforms.com/` to the link.

stephywells replied:
Oh, right. How about we use it for any links to our site? The links to their site don't actually need it yet since we don't have any related usage tracking set up to save clicks to their site and send them to us.

truongwp replied:
Ok. So I will remove tracking params from user site URLs.

stephywells commented:
>  For the Offer help part, what is the link for the Contact support? Is it https://formidableforms.com/knowledgebase/support-2/?

Let's use https://formidableforms.com/new-topic/ for the support url

mikeldalton commented:
This is the text for the expired email:

I hope you have loved the past year of using Formidable Forms! Your subscription was set to manually renew and has just expired.
Renewing your license grants you access to our legendary support services, form templates, Pro features, and updates for another year.
If you don't need Formidable Pro right now, that's OK. You can let this subscription expire and simply resubscribe in the future when needed.

truongwp commented:
@mikeldalton Thank you. I updated it. Once you get the latest code of this PR, you will need to get this Pro version to test the license expired email: https://github.com/Strategy11/formidable-pro/archive/refs/heads/license-expired-email.zip

To test it, run `FrmProEmailLicenseExpiredController::send()`.

stephywells reviewed (changes requested):
This is great @truongwp ! These emails look awesome. Thanks for making these updates!

I just have a couple more small questions, and it looks like there are a few updates that need to be made for the tests that aren't passing. This is so close to ready!

truongwp commented on classes/helpers/FrmEmailSummaryHelper.php:
```diff
@@ -0,0 +1,516 @@
+		// If renewal date doesn't exist, return 365 days since the first form.
+		$first_form_date = self::get_earliest_form_created_date();
+		if ( $first_form_date ) {
+			$renewal_date = gmdate( 'Y-m-d', strtotime( $first_form_date . '+' . self::YEARLY_PERIOD . ' days' ) );
```
It won't work. I updated it to set the renewal date to the next 45 days in this case.

truongwp replied:
Yes, we can. I updated it to the next 45 days from today to send the yearly email as soon as possible after updating Formidable Forms.

stephywells commented on classes/views/summary-emails/stats.php:
```diff
@@ -0,0 +1,130 @@
+	</div>
+
+	<!-- Top forms section -->
+	<div style="<?php echo esc_attr( FrmEmailSummaryHelper::get_section_style() ); ?>">
```
I have an empty section here in my latest test email. Let's skip this section if top_forms is empty.

truongwp replied:
Thank you. I updated it. Please check again.

stephywells reviewed (approved):
This looks great!!

I just came across one last thing. Then we can merge this!

Crabcyborg commented on classes/helpers/FrmEmailSummaryHelper.php:
```diff
@@ -0,0 +1,529 @@
+			array(
+				'created_at >' => $from_date,
+				'created_at <' => $to_date . ' 23:59:59',
+				'status'       => 'complete',
```
@truongwp In newer versions of Stripe Lite we now have a `test` column.

We could check if the `frm_trans_db_version` option from `FrmTransLiteDb` is up to date (version 6).

If it is, we could query where `'test' => 0`.

If `test` is `null`, it's old data, and we don't know if it's `test` or `live`. If it's 1, it's `test` and if it's 0, it's `live`.

Crabcyborg commented on classes/helpers/FrmEmailSummaryHelper.php:
```diff
@@ -0,0 +1,529 @@
+			'frm_items',
+			array(
+				'created_at >' => $from_date, // The `=` is added after `>` in the query.
+				'created_at <' => $to_date . ' 23:59:59',
```
Should these ranges be localized to the user's timezone?

Crabcyborg commented on classes/helpers/FrmEmailSummaryHelper.php:
```diff
@@ -0,0 +1,529 @@
+						WHERE it.created_at BETWEEN %s AND %s AND it.is_draft = 0
+						GROUP BY form_id ORDER BY items_count DESC LIMIT %d",
+				$from_date,
+				$to_date . ' 23:59:59',
```
I'm curious about timezones here as well.

This is always UTC, right?

Could we localize it?

truongwp replied:
I'm not sure if we need the local timezone or not. Maybe a bit of statistical data goes to the next or previous month, which I think it's acceptable. @stephywells What do you think about this?

stephywells replied:
I don't think it's worth a lot of time, but I think it would be quick so we might as well. I think this should do the job:
```suggestion
				FrmAppHelper::get_localized_date( 'Y-m-d', $to_date . ' 23:59:59' ) ,
```

stephywells replied:
No.... that's not what you're saying. Yes! That would make more sense.

Crabcyborg commented on classes/controllers/FrmHooksController.php:
```diff
@@ -107,6 +107,12 @@ public static function load_hooks() {
 		add_filter( 'frm_fields_in_form_builder', 'FrmFormsController::update_form_builder_fields', 10, 2 );
 
+		// Cronjob.
+		add_action( 'admin_init', 'FrmCronController::schedule_events' );
```
@truongwp Can we put this hook in `load_admin_hooks`?

I think we can also put the `add_action( 'frm_daily_event', 'FrmEmailSummaryController::maybe_send_emails' );` line under this in `load_admin_hooks`, right?

truongwp replied:
I moved this line to the `load_admin_hooks()`. But I don't think we should move the `frm_daily_event` line there.

Crabcyborg commented on classes/helpers/FrmEmailSummaryHelper.php:
```diff
@@ -0,0 +1,529 @@
+			array(
+				'created_at >' => $from_date,
+				'created_at <' => $to_date . ' 23:59:59',
+				'status'       => 'complete',
```
The `status` column is a bit problematic.

If people are only using the PayPal add on, it's possible there is no `status` column in the payments table. Instead they have a `completed` column.

So if PayPal is active, and the Lite migrations have not run, we'll want to query for `'completed' => 1` instead.

truongwp replied:
I don't think there is any chance this migration hasn't run: https://github.com/Strategy11/formidable-forms/blob/39cefce8d02447ec04c55c554ed540f871e200a8/stripe/models/FrmTransLiteDb.php#L301

If it hasn't run before (because of using an old Formidable Forms version with Paypal addon only), it will run after updating Formidable Forms to this version.

Crabcyborg replied:
@truongwp The migration methods only run from `FrmStrpLiteConnectHelper::set_stripe_details_as_submitted` though. It happens on Stripe connection, not when the plugin updates.

I've already made a couple updates to work better with the PayPal add on when Stripe is not connected.

- https://github.com/Strategy11/formidable-forms/pull/1380
- https://github.com/Strategy11/formidable-forms/pull/1384

It's fairly easy to test. You just need to delete your payments table, subscriptions table, the `frm_trans_db_version` and `frm_pay_db_version` options, disable the stripe add on, and activate the PayPal add on.

truongwp commented:
@Crabcyborg @stephywells I updated the code to use the localized date. Any date we calculate from "today date" will need to run through the localized date method. Any date we calculate from a date string won't need to run through it.

I also changed a bit in the `FrmEmailStats` class to reduce duplicated code in its children.

truongwp commented:
@stephywells Will we hide the Dashboard link in the email until the Dashboard UI released?

stephywells commented:
> @stephywells Will we hide the Dashboard link in the email until the Dashboard UI released?

We can keep the link. Right now it's fine to go to the form listing page.

Crabcyborg commented:
@garretlaxton I think we're close to merging this.

Could you go back through and do some more QA first?

Thank you!

Crabcyborg commented on classes/helpers/FrmEmailSummaryHelper.php:
```diff
@@ -0,0 +1,567 @@
+	 * @param string $to_date   To date.
+	 * @return array            Contains `count` and `total`.
+	 */
+	public static function get_payments_data( $from_date, $to_date ) {
```
Can we possibly make this function more universal? Liviu is working on the same sort of query for the dashboard and it would be nice if he could use this, but not as part of the email summary helper.

truongwp replied:
@Crabcyborg Is it ok to create a `FrmStatsHelper` class?

Crabcyborg replied:
That sounds finer.
Maybe it make sense in FrmTransLiteDb.php though?
The other payments db queries are in there so it would be consistent there.

truongwp replied:
That makes sense. I moved it. Please check again.

Crabcyborg reviewed (approved):
Thank you @truongwp!

🚀

Crabcyborg commented:
I'm going to go ahead and merge this so we can try to use the payment code in the dashboard plugin.

If anyone finds any issues, feel free to make a new issue in Pro.

stephywells commented on classes/views/summary-emails/base-plain.php:
```diff
@@ -0,0 +1,25 @@
+$frm_settings = FrmAppHelper::get_settings();
+
+echo '%%INNER_CONTENT%%';
+
```
We could use a line break here. We could remove the extra from the bottom of "out of date plugins" in the other template. Right now, there is no break in the renewal email.

truongwp replied:
If there are no out-of-date plugins, we will have extra line breaks from the top forms section. We can use str_replace() to add the line breaks for the license expired email.

stephywells replied:
That works too. Thank you!

stephywells commented on classes/views/summary-emails/base-plain.php:
```diff
@@ -0,0 +1,25 @@
+echo '%%INNER_CONTENT%%';
+
+// translators: contact support URL.
+printf( esc_html__( 'Need help? Get in touch with our team at %s', 'formidable' ), esc_url( $args['support_url'] ) );
```
The plain text urls (support and unsubscribe) aren't working right. It solves the problem to switch to esc_url_raw.

stephywells reviewed (commented):
@truongwp I ran into a couple plain text issues while testing the renewal email. Would you mind creating a new PR for these fixes?
